apps/web/src/hooks/useRegisterNameCallback.ts

```diff
@@ -85,6 +99,35 @@ export function useRegisterNameCallback(
   const normalizedName = normalizeEnsDomainName(name);
   const isDiscounted = Boolean(discountKey && validationData);
 
+  const prepareReverseRecordSignature = useCallback(async () => {
+    if (!address) throw new Error('No address');
+    if (!publicClient) throw new Error('No public client');
+
+    // Use the L2 Reverse Registrar address configured for this chain.
+    const reverseRegistrar = USERNAME_L2_REVERSE_REGISTRAR_ADDRESSES[basenameChain.id];
+
+    const functionAbi = L2ReverseRegistrarAbi.find(
+      (f) => f.type === 'function' && f.name === 'setNameForAddrWithSignature',
+    ) as unknown as AbiFunction;
+    const selector = getFunctionSelector(functionAbi);
+
+    const signatureExpiry = BigInt(Math.floor(Date.now() / 1000) + 5 * 60);
+    const coinTypes = [BigInt(convertChainIdToCoinTypeUint(basenameChain.id))] as const;
+    const fullName = formatBaseEthDomain(name, basenameChain.id);
+
+    const preimage = encodePacked(
+      ['address', 'bytes4', 'address', 'uint256', 'string', 'uint256[]'],
+      [reverseRegistrar, selector, address, signatureExpiry, fullName, coinTypes],
+    );
+
+    const digest = keccak256(preimage);
```

Comment:
I think it would be cleaner to move some of this logic to a separate function in another file

Reply:
yeah good idea, especially since we'll need the same sig in the setPrimary flow later